src/imp/mbedtls.rs

```diff
@@ -0,0 +1,522 @@
+    Der(pkcs5::der::Error),
+    Custom(String),
+}
+
+impl From<TlsError> for Error {
+    fn from(err: TlsError) -> Error {
+        Error::Tls(err)
+    }
+}
+
+impl From<yasna::ASN1Error> for Error {
+    fn from(err: yasna::ASN1Error) -> Error {
+        Error::Pkcs12(err)
+    }
+}
+
+impl From<pkcs5::Error> for Error {
+    fn from(err: pkcs5::Error) -> Error {
+        Error::Pkcs5(err)
+    }
+}
+
+impl From<pkcs5::der::Error> for Error {
+    fn from(err: pkcs5::der::Error) -> Error {
+        Error::Der(err)
+    }
+}
+
+impl<S> From<TlsError> for HandshakeError<S> {
+    fn from(e: TlsError) -> HandshakeError<S> {
+        HandshakeError::Failure(e.into())
+    }
+}
+
+impl error::Error for Error {
+    fn source(&self) -> Option<&(dyn error::Error + 'static)> {
+        match *self {
+            Error::Tls(ref e) => e.source(),
+            Error::Pkcs12(ref e) => e.source(),
+            Error::Pkcs5(_) => None,
+            Error::Der(_) => None,
+            Error::Custom(_) => None,
+        }
+    }
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            Error::Tls(ref e) => fmt::Display::fmt(e, fmt),
+            Error::Pkcs12(ref e) => fmt::Display::fmt(e, fmt),
+            Error::Pkcs5(ref e) => fmt::Display::fmt(e, fmt),
+            Error::Der(ref e) => fmt::Display::fmt(e, fmt),
+            Error::Custom(ref e) => fmt::Display::fmt(e, fmt),
+        }
+    }
+}
+
+fn to_mbedtls_version(protocol: Protocol) -> Version {
+    match protocol {
+        Protocol::Sslv3 => Version::Ssl3,
+        Protocol::Tlsv10 => Version::Tls1_0,
+        Protocol::Tlsv11 => Version::Tls1_1,
+        Protocol::Tlsv12 => Version::Tls1_2,
+    }
+}
+
+trait NullTerminated {
+    fn null_terminated(&self) -> Vec<u8>;
+}
+
+impl<T: AsRef<[u8]>> NullTerminated for T {
+    fn null_terminated(&self) -> Vec<u8> {
+        let mut buf = self.as_ref().to_vec();
+        buf.push(0);
+        buf
+    }
+}
+
+fn pkcs12_decode_key_bag<B: AsRef<[u8]>>(
+    key_bag: &p12::EncryptedPrivateKeyInfo,
+    pass: B,
+) -> Result<Vec<u8>, Error> {
+    // try to decrypt the key with algorithms supported by p12 crate
+    if let Some(decrypted) = key_bag.decrypt(pass.as_ref()) {
+        Ok(decrypted)
+    // try to decrypt the key with algorithms supported by pkcs5 standard
+    } else if let p12::AlgorithmIdentifier::OtherAlg(_) = key_bag.encryption_algorithm {
+        // write the algorithm identifier back to DER format
+        let algorithm_der =
+            yasna::construct_der(|writer| key_bag.encryption_algorithm.write(writer));
+        // and construct pkcs5 decoder from it
+        let scheme = pkcs5::EncryptionScheme::try_from(&algorithm_der[..])?;
+
+        Ok(scheme.decrypt(pass.as_ref(), &key_bag.encrypted_data)?)
+    } else {
+        Err(Error::Custom(
+            "Unsupported key encryption algorithm".to_owned(),
+        ))
+    }
+}
```

Comment:
Custom pkcs12 treatment is there only because `rust-mbedtls` implementation doesn't support pbes2 encrypted keys (which is openssl's default). I'll try to provide support for it directly in mbedtls instead

Reply:
@sfackler @jethrogb
I've checked and we could add an implementation of pbes2 to rust-mbedtls, however from encryption schemes defined by [pkcs5 rfc](https://datatracker.ietf.org/doc/html/rfc8018#appendix-B.2), MbedTls [has support](https://github.com/fortanix/rust-mbedtls/blob/master/mbedtls-sys/vendor/library/oid.c#L550) only for des-cbc and des-ede3-cbc.
rc2-cbc, rc5-cbc and aes-cbc are missing and aes-cbc is openssl's default when running `openssl pkcs12 -export`.
rc2 and rc5 are also missing in `pkcs5` crate, so it seems to be all about AES.
Moreover, des-cbc is considered insecure: 'Although its short key length of 56 bits makes it too insecure for modern applications' [wiki](https://en.wikipedia.org/wiki/Data_Encryption_Standard)

Here are the options:
 - Add native (MbedTls) support for pbes2 in `rust-mbedtls` and use it there, only allowing des-cbc and des-ede3-cbc encryption schemes and providing compatible `.p12` for `rust-native-tls` tests
 - Add non-native (using pkcs5 crate) in `rust-mbedtls` and maybe hide it behind feature flag, allowing also aes-cbc
 - Keep this implementation and don't use `rust-mbedtls` for pbes2 (however if I'll have some spare time I'll likely issue a PR for `rust-mbedtls` anyway as I've already written most of the implementation)

Reply:
https://github.com/Mbed-TLS/mbedtls/issues/7024
https://github.com/Mbed-TLS/mbedtls/issues/7038
It's possible AES will get support in MbedTls at some point, but it'll probably require rust-mbedtls to update to 3.x then, which can be major effort

Reply:
> it'll probably require rust-mbedtls to update to 3.x

This is actually ongoing

Reply:
The preferred way of solving this would be to add support for AES in `MbedTls`: https://github.com/Mbed-TLS/mbedtls/pull/7604
then pull it into `rust-mbedtls` and use it directly here

Reply:
we've been discussing this with @jethrogb and it seems that `rust-mbedtls` is rather going to take steps towards removing pkcs12 support than extending it to pbes2. So it is up to you @sfackler whether we leave this implemented as is now using p12 and pkcs5 crates or just mark it as unimplemented.